### PR TITLE
Implement Object Detection Training

### PR DESCRIPTION
**Pull request by liopeer**
## What has changed and why?
- adds object detection training

## How has it been tested?
- locally

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [ ] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [ ] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_data/task_batch_collation.py [hunk 1]
```diff
@@ -71,15 +72,22 @@ def __init__(
...
-            self.scale_jitter = ScaleJitter(
-                sizes=transform_args.scale_jitter.sizes,
-                target_size=transform_args.image_size,
-                scale_range=scale_range,
-                num_scales=transform_args.scale_jitter.num_scales,
-                divisible_by=transform_args.scale_jitter.divisible_by,
-                p=transform_args.scale_jitter.prob,
-                step_seeding=transform_args.scale_jitter.step_seeding,
-                seed_offset=transform_args.scale_jitter.seed_offset,
+            self.scale_jitter = Compose(
```
Comment (liopeer) [inline src/lightly_train/_data/task_batch_collation.py:97]: Has to be wrapped in `Compose` order to work correctly on the Bboxes.

File: src/lightly_train/_data/task_batch_collation.py
Comment (liopeer) [inline src/lightly_train/_data/task_batch_collation.py:137]: Moved this during debugging. Could also revert.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/train_model.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#
+# Copyright (c) Lightly AG and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+#
+from typing import ClassVar, Literal
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov3_ltdetr_object_detection/train_model.py]: This will follow. Already big enough as it is right now.

File: src/lightly_train/_task_models/object_detection_components/hybrid_encoder.py [hunk 3]
```diff
@@ -303,9 +303,11 @@ def __init__(
         self.downsample_convs = nn.ModuleList()
         self.pan_blocks = nn.ModuleList()
         for _ in range(len(in_channels) - 1):
-            self.downsample_convs.append(
-                ConvNormLayer(hidden_dim, hidden_dim, 3, 2, act=act)
-            )
+            # Only allocate if needed, otherwise causes errors in distributed training.
```
Comment (liopeer) [inline src/lightly_train/_task_models/object_detection_components/hybrid_encoder.py]: TODO: Add this change to the `NOTICE` file and make comment more clear. Training fails because this is detected as an unused parameter.

File: src/lightly_train/_task_models/object_detection_components/rtdetrv2_criterion.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+import torchvision
+
+from lightly_train._task_models.object_detection_components.box_ops import (
+    box_cxcywh_to_xyxy,
+    box_iou,
+    generalized_box_iou,
+)
+
+
+def is_dist_available_and_initialized():
```
Comment (liopeer) [inline src/lightly_train/_task_models/object_detection_components/rtdetrv2_criterion.py:30]: This function is not from the original file. TODO: Add comment and to `NOTICE` file.

File: src/lightly_train/_task_models/object_detection_components/rtdetrv2_criterion.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    def get_loss(self, loss, outputs, targets, indices, num_boxes, **kwargs):
+        loss_map = {
+            "boxes": self.loss_boxes,
+            "focal": self.loss_labels_focal,
+            "vfl": self.loss_labels_vfl,
+        }
+        assert loss in loss_map, f"do you really want to compute {loss} loss?"
+        return loss_map[loss](outputs, targets, indices, num_boxes, **kwargs)
+
+    def forward(self, outputs, targets, world_size, **kwargs):
```
Comment (liopeer) [inline src/lightly_train/_task_models/object_detection_components/rtdetrv2_criterion.py:188]: `world_size` added by me, so that we can use Lightning Fabric's world_size. TODO: Add change to `NOTICE` and add comment.
Rest of the file is original.

File: src/lightly_train/_transforms/object_detection_transform.py [hunk 6]
```diff
@@ -169,6 +155,14 @@ def __init__(
                     VerticalFlip(p=transform_args.random_flip.vertical_prob)
                 ]
 
+        if transform_args.resize is not None:
```
Comment (liopeer) [inline src/lightly_train/_transforms/object_detection_transform.py:174]: For the ValTransform we use Resize and skip the ScaleJitter in the collate function.

File: src/lightly_train/_commands/train_task.py [hunk 7]
```diff
@@ -37,6 +40,135 @@
...
+    precision: _PRECISION_INPUT = "bf16-mixed",
+    float32_matmul_precision: Literal["auto", "highest", "high", "medium"] = "auto",
+    seed: int | None = 0,
+    logger_args: dict[str, Any] | None = None,
+    model_args: dict[str, Any] | None = None,
+    transform_args: dict[str, Any] | None = None,
+    loader_args: dict[str, Any] | None = None,
+    save_checkpoint_args: dict[str, Any] | None = None,
+) -> None:
+    """Train a semantic segmentation model.
```
Comment (stegmuel) [inline src/lightly_train/_commands/train_task.py]: Can you remove semantic segmentation references.

File: src/lightly_train/_data/yolo_object_detection_dataset.py [hunk 8]
```diff
@@ -167,7 +171,23 @@ def _replace_first_images_with_labels(path: Path) -> Path:
...
     label_dir: Path
     classes: dict[int, str]
+
+    def list_image_info(self) -> Iterable[dict[str, str]]:
+        for image_filename in file_helpers.list_image_filenames_from_dir(
+            image_dir=self.image_dir
+        ):
+            image_filepath = self.image_dir / Path(image_filename)
+            label_filepath = self.label_dir / Path(image_filename).with_suffix(".txt")
+            if label_filepath.exists():
```
Comment (stegmuel) [inline src/lightly_train/_data/yolo_object_detection_dataset.py:160]: Can you add a TODO to log a message when image-label pairs are not complete?

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        self.model = DINOv2LTDetrObjectDetection(
+            model_name=model_name,
+            image_size=val_transform_args.image_size,
+            classes=data_args.names,
+            image_normalize=None,  # TODO (Lionel, 10/25): Allow custom normalization.
+            backbone_weights=model_args.backbone_weights,
+            backbone_args=model_args.backbone_args,  # TODO (Lionel, 10/25): Potentially remove in accordance with EoMT.
+        )
+
+        matcher = HungarianMatcher(  # type: ignore[no-untyped-call]
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:152]: Should we add a TODO to have a HungarianMatcherArgs class?

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+            weight_dict={"cost_class": 2, "cost_bbox": 5, "cost_giou": 2},
+            use_focal_loss=True,
+            alpha=0.25,
+            gamma=2.0,
+        )
+
+        self.criterion = RTDETRCriterionv2(  # type: ignore[no-untyped-call]
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:159]: Should we add a TODO to have a CriterionArgs class?

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+                "name": "detector",
+                "params": [
+                    p
+                    for n, p in self.model.named_parameters()
+                    if re.match(r"^(?=.*(?:encoder|decoder))(?=.*(?:norm|bn)).*$", n)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        optim = AdamW(
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:307]: The possibility to customize the optimizer args will follow in an other pr?

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+            loss_dict = self.criterion(
+                outputs=outputs,
+                targets=targets,
+                epoch=None,
+                step=None,
+                global_step=None,
+                world_size=fabric.world_size,
+            )
+
+        total_loss = sum(loss_dict.values())
```
Comment (stegmuel) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:252]: Are the losses aggregated across GPUs somewhere ?

Comment (liopeer) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:252]: Good question. I though Fabric was doing this automatically, but I will have to check.

File: src/lightly_train/_commands/train_task_helpers.py [hunk 13]
```diff
@@ -45,6 +42,9 @@
 from lightly_train._task_models.dinov2_linear_semantic_segmentation.train_model import (
     DINOv2LinearSemanticSegmentationTrain,
 )
+from lightly_train._task_models.dinov2_ltdetr_object_detection.train_model import (
+    DINOv2LTDetrObjectDetectionTrain,
```
Comment (guarin) [inline src/lightly_train/_commands/train_task_helpers.py]:
```suggestion
    DINOv2LTDETRObjectDetectionTrain,
```

AFAIK DETR is always spelled with all caps

File: src/lightly_train/_commands/train_task_helpers.py [hunk 14]
```diff
@@ -492,15 +490,15 @@ def get_dataset_mmap_file(
...
+    dataset_args: TaskDatasetArgs,
     transform: TaskTransform,
     mmap_filepath: Path,
 ) -> TaskDataset:
     image_info = dataset_args.list_image_info()
 
     dataset_cls = dataset_args.get_dataset_cls()
     # TODO(Guarin, 08/25): Relax this when we add object detection.
-    assert isinstance(transform, SemanticSegmentationTransform)
+    assert isinstance(transform, TaskTransform)
```
Comment (guarin) [inline src/lightly_train/_commands/train_task_helpers.py]:
```suggestion
```

Isn't needed anymore as it now matches the type in the function definition

File: src/lightly_train/_data/mask_semantic_segmentation_dataset.py [hunk 15]
```diff
@@ -107,18 +108,19 @@ def is_mask_valid(self, mask: Tensor) -> bool:
...
                 class_id
-                for class_id in self.args.classes.keys()
+                for class_id in self.dataset_args.classes.keys()
                 if class_id not in ignore_classes
             )
         }
 
     def __len__(self) -> int:
-        return len(self.filepaths)
+        return len(self.image_info)
```
Comment (guarin) [inline src/lightly_train/_data/mask_semantic_segmentation_dataset.py]: `__len__` can be moved to superclass if `image_info` is saved there

File: docs/prebuild.py [hunk 16]
```diff
@@ -94,7 +100,11 @@ def dump_transform_args_for_methods(dest_dir: Path) -> None:
 def dump_transform_args_for_tasks(dest_dir: Path) -> None:
     dest_dir.mkdir(parents=True, exist_ok=True)
     for train_model_cls in TASK_TRAIN_MODEL_CLASSES:
-        if train_model_cls in {DINOv2LinearSemanticSegmentationTrain}:
+        if train_model_cls in {
+            DINOv2LinearSemanticSegmentationTrain,
+            DINOv2LTDETRObjectDetectionTrain,
```
Comment (liopeer) [inline docs/prebuild.py:106]: We should still add those, right? Can we add a TODO?

File: src/lightly_train/_commands/train_task.py [hunk 17]
```diff
@@ -547,11 +687,21 @@ def _train_task_from_config(config: TrainTaskConfig) -> None:
...
 
 
+def _is_better_metric(
+    current_metric: float, best_metric: float, mode: Literal["min", "max"]
+) -> bool:
+    if mode == "min":
+        return current_metric < best_metric
+    elif mode == "max":
+        return current_metric > best_metric
+    raise ValueError(f"Unknown mode: {mode}")
```
[the remark this answers is not in the document]
Comment (liopeer) [reply src/lightly_train/_commands/train_task.py:730]: Yes, that's indeed something to address!

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py [hunk 18]
```diff
@@ -37,7 +37,7 @@
 logger = logging.getLogger(__name__)
 
 
-class DINOv2LTDetrObjectDetectionTaskModel(TaskModel):
+class DINOv2LTDETRObjectDetection(TaskModel):
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:40]: I am fine with renaming this, but that means that we will have to quickly change the checkpoints as well (little effort though).

File: src/lightly_train/_transforms/random_iou_crop.py [hunk 19]
```diff
@@ -177,7 +177,7 @@ def get_params_dependent_on_data(
             return {"crop_coords": crop_coords, "pad_params": None}
         else:
             return {
-                "crop_coords": (0, 0, orig_image_shape[0], orig_image_shape[1]),
+                "crop_coords": (0, 0, orig_image_shape[1], orig_image_shape[0]),
```
Comment (liopeer) [inline src/lightly_train/_transforms/random_iou_crop.py:180]: Oh damn, that's a good catch! I had tested this thoroughly but maybe only with square images where this was not caught.

File: src/lightly_train/_distributed.py [hunk 20]
```diff
@@ -52,3 +56,32 @@ def is_node_rank_zero() -> bool:
...
+        # Ensure all values are tensors on the same device
+        values = [
+            v.detach().to(torch.device("cuda" if torch.cuda.is_available() else "cpu"))
+            for v in values
+        ]
+        stacked = torch.stack(values, dim=0)
+
+        dist.all_reduce(stacked)
+
+        if average:
```
Comment (liopeer) [inline src/lightly_train/_distributed.py:83]: Nit: Probably more efficient if you use
```python
from torch.distributed import ReduceOp

dist.all_reduce(stacked, op=ReduceOp.AVG if average else ReduceOp.SUM)
```
I think it then does everything directly via CUDA/NCCL and does not return to PyTorch.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+
+    save_checkpoint_args_cls: ClassVar[type[TaskSaveCheckpointArgs]] = (
+        DINOv2LTDETRObjectDetectionTaskSaveCheckpointArgs
+    )
+
+    backbone_weights: PathLike | None = None
+    backbone_url: str = ""
+    backbone_args: dict[str, Any] = {}
+
+    # Matcher configuration
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:78]: I would ultimately prefer to have `MatcherArgs`, `CriterionArgs` and `OptimizerArgs` classes, but we can leave it like this for now.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+                "weight_decay": self.model_args.detector_weight_decay,
+            },
+        ]
+        optim = AdamW(
+            param_groups,
+            lr=self.model_args.optimizer_lr,
+            betas=self.model_args.optimizer_betas,
+            weight_decay=self.model_args.optimizer_weight_decay,
+        )
+        scheduler = MultiStepLR(
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:313]: Will you now change this to cosine with warmup?

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/train_model.py [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+                "weight_decay": self.model_args.detector_weight_decay,
+            },
+        ]
+        optim = AdamW(
+            param_groups,
+            lr=self.model_args.optimizer_lr,
+            betas=self.model_args.optimizer_betas,
+            weight_decay=self.model_args.optimizer_weight_decay,
+        )
+        scheduler = MultiStepLR(
```
Comment (liopeer) [inline src/lightly_train/_task_models/dinov3_ltdetr_object_detection/train_model.py:312]: Same here.

File: docs/prebuild.py (hunk 16 above)
Comment (stegmuel) [reply docs/prebuild.py:106]: Yes, once we have the doc ready. I left a TODO.

File: src/lightly_train/_distributed.py (hunk 20 above)
Comment (stegmuel) [reply src/lightly_train/_distributed.py:83]: Supposedly, ReduceOp.AVG is not implemented for the nccl backend.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py (hunk 18 above)
Comment (stegmuel) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/task_model.py:40]: Yes...

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py (hunk 21 above)
Comment (stegmuel) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:78]: Left a TODO. We also have everything in one place for EoMT.

File: src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py (hunk 22 above)
Comment (stegmuel) [reply src/lightly_train/_task_models/dinov2_ltdetr_object_detection/train_model.py:313]: I'll do this in the small model PR yes.

File: src/lightly_train/_task_models/dinov3_ltdetr_object_detection/train_model.py (hunk 23 above)
Comment (stegmuel) [reply src/lightly_train/_task_models/dinov3_ltdetr_object_detection/train_model.py:312]: Same answer.

File: src/lightly_train/_commands/train_task.py [hunk 24]
```diff
@@ -418,7 +552,7 @@ def _train_task_from_config(config: TrainTaskConfig) -> None:
             logger.info(f"Training for {config.steps} steps...")
 
         fabric.barrier()
-        max_val_miou = 0.0
+        best_metric = 0.0 if config.save_checkpoint_args.mode == "max" else float("inf")
```
Comment (liopeer) [inline src/lightly_train/_commands/train_task.py]:
```suggestion
        best_metric = -float("inf") if config.save_checkpoint_args.mode == "max" else float("inf")
```
Theoretically we could have a metric e.g. converging to 0 from the left side.
